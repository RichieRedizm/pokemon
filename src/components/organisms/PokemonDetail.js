import React, { Fragment, useEffect, useContext } from 'react'
import { Link } from 'react-router-dom'
import GraphQLContext from '../../context/graphQL/graphQLContext'
import Spinner from '../search/Spinner'
import Attacks from '../molecules/PokemonAttacks'
const { formatArrayList, formatMinMaxList } = require('../helper')

const PokemonDetail = ({ match }) => {
  const graphQLContext = useContext(GraphQLContext)
  const { loading, pokemon, pokemonDetails } = graphQLContext

  useEffect(() => {
    pokemonDetails(match.params.id, match.params.name)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  if (loading || pokemon === null) return <Spinner />

  const {
    name,
    image,
    classification,
    types,
    weight,
    height,
    resistant,
    weaknesses,
    attacks
  } = pokemon

  // TODO: very bloated - need to split into atoms
  return (
    <Fragment>
      <Link to={`/`} className='btn btn-light btn-sm'>
        Back to Pokemon list
      </Link>
      <div className='detail-wrap' data-testid='pokemon-detail'>
        <div className='detail-meta'>
          {name && <h2>{name}</h2>}
          {classification && <h3 className='label-dark'>{classification}</h3>}
        </div>
        <div className='detail-image'>
          <img src={image} alt={name} />
        </div>
      </div>
      <div className='grid-2'>
        <div className='card detail'>
          <h3 className='label-dark'>Statistics</h3>
          {types && (
            <div>
              <strong className='label-dark'>Type: </strong>
              {formatArrayList(types)}
            </div>
          )}
          {weight && (
            <div>
              <strong className='label-dark'>Weight: </strong>
              {formatMinMaxList(weight)}
            </div>
          )}
          {height && (
            <div>
              <strong className='label-dark'>Height: </strong>
              {formatMinMaxList(height)}
            </div>
          )}
          {resistant && (
            <div>
              <strong className='label-dark'>Resistant: </strong>
              {formatArrayList(resistant)}
            </div>
          )}
          {weaknesses && (
            <div>
              <strong className='label-dark'>weaknesses: </strong>
              {formatArrayList(weaknesses)}
            </div>
          )}
        </div>
        {attacks && <Attacks attacks={attacks} />}
      </div>
    </Fragment>
  )
}

export default PokemonDetail
