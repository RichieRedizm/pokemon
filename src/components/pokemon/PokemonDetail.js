import React, { Fragment, useEffect, useContext } from 'react'
import { Link } from 'react-router-dom'
import GraphQLContext from '../../context/graphQL/graphQLContext'
import Spinner from '../search/Spinner'
const {
  formatArrayList,
  formatMinMaxList,
  formatObjectList
} = require('../helper')

const PokemonDetail = ({ match }) => {
  const graphQLContext = useContext(GraphQLContext)
  const { loading, pokemon, pokemonDetails } = graphQLContext

  useEffect(() => {
    pokemonDetails(match.params.id, match.params.name)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  if (loading || pokemon === null) return <Spinner />

  const {
    name,
    image,
    classification,
    types,
    weight,
    height,
    resistant,
    weaknesses,
    attacks
  } = pokemon

  return (
    <Fragment>
      <Link to={`/`} className='btn btn-light btn-sm'>
        Back to Pokemon list
      </Link>
      <div className='detail-wrap'>
        <div className='detail-meta'>
          {name && <h2>{name}</h2>}
          {classification && <h3 className='label-dark'>{classification}</h3>}
        </div>
        <div className='detail-image'>
          <img src={image} alt={name} />
        </div>
      </div>
      <div className='grid-2'>
        <div className='card detail'>
          {types && <strong className='label-dark'>Type:</strong>}
          {types && formatArrayList(types)}
          {weight && <strong className='label-dark'>Weight:</strong>}
          {weight && formatMinMaxList(weight)}
          {height && <strong className='label-dark'>Weight:</strong>}
          {height && formatMinMaxList(height)}
          {resistant && <strong className='label-dark'>Resistant: </strong>}
          {resistant && formatArrayList(resistant)}
          {weaknesses && <strong className='label-dark'>weaknesses: </strong>}
          {weaknesses && formatArrayList(weaknesses)}
        </div>
        <div className='card detail'>
          {attacks.fast && (
            <strong className='label-dark'>Fast Attacks: </strong>
          )}
          {attacks.fast && formatObjectList(attacks.fast)}
          {attacks.special && (
            <strong className='label-dark'>Special Attacks: </strong>
          )}
          {attacks.special && formatObjectList(attacks.special)}
        </div>
      </div>
    </Fragment>
  )
}

export default PokemonDetail
