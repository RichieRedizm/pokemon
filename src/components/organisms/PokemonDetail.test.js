import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import GraphQLContext from '../../context/graphQL/graphQLContext'
import { render, cleanup } from '@testing-library/react'
import PokemonDetail from './PokemonDetail'

afterEach(cleanup)

function renderPokemonDetail(state) {
  const match = {
    params: {
      id: 'UG9rZW1vbjowMDE=',
      name: 'Bulbasaur'
    }
  }
  return render(
    <GraphQLContext.Provider value={state}>
      <Router>
        <PokemonDetail match={match} />
      </Router>
    </GraphQLContext.Provider>
  )
}
test('if loading set to true load Spinner component', () => {
  const { getByTestId, asFragment } = renderPokemonDetail({
    pokemonDetails: jest.fn(),
    pokemons: [],
    loading: true
  })

  expect(getByTestId('loading'))
  expect(asFragment()).toMatchSnapshot()
})

test('if pokemon data not null generate statistics detail', () => {
  const pokemonDetailData = {
    pokemonDetails: jest.fn(),
    pokemon: {
      name: 'Bulbasaur',
      weight: {
        minimum: '6.04kg',
        maximum: '7.76kg'
      },
      height: {
        minimum: '0.61m',
        maximum: '0.79m'
      },
      classification: 'Seed Pok√©mon',
      types: ['Grass', 'Poison'],
      resistant: ['Water', 'Electric', 'Grass', 'Fighting', 'Fairy'],
      weaknesses: ['Fire', 'Ice', 'Flying', 'Psychic'],
      attacks: {
        fast: [
          {
            name: 'Tackle',
            type: 'Normal',
            damage: 12
          },
          {
            name: 'Vine Whip',
            type: 'Grass',
            damage: 7
          }
        ],
        special: [
          {
            name: 'Power Whip',
            type: 'Grass',
            damage: 70
          },
          {
            name: 'Seed Bomb',
            type: 'Grass',
            damage: 40
          },
          {
            name: 'Sludge Bomb',
            type: 'Poison',
            damage: 55
          }
        ]
      },
      image: 'https://img.pokemondb.net/artwork/bulbasaur.jpg'
    },
    loading: false
  }

  const { getByTestId, asFragment } = renderPokemonDetail(pokemonDetailData)
  expect(getByTestId('pokemon-detail'))
  expect(asFragment()).toMatchSnapshot()
})
