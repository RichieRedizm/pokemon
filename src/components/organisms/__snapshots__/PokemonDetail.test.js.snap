// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if loading set to true load Spinner component 1`] = `
<DocumentFragment>
  <img
    alt="loading"
    data-testid="loading"
    src="spinner.gif"
    style="width: 200px; margin: auto; display: block;"
  />
</DocumentFragment>
`;

exports[`if pokemon data not null generate statistics detail 1`] = `
<DocumentFragment>
  <a
    class="btn btn-light btn-sm"
    href="/"
  >
    Back to Pokemon list
  </a>
  <div
    class="detail-wrap"
    data-testid="pokemon-detail"
  >
    <div
      class="detail-meta"
    >
      <h2>
        Bulbasaur - 001
      </h2>
      <h3
        class="label-dark"
      >
        Seed Pok√©mon
      </h3>
    </div>
    <div
      class="detail-image"
    >
      <img
        alt="Bulbasaur"
        src="https://img.pokemondb.net/artwork/bulbasaur.jpg"
      />
    </div>
  </div>
  <div
    class="grid-12"
  >
    <div
      class="card detail"
    >
      <h3
        class="label-dark"
      >
        Statistics
      </h3>
      <div>
        <strong
          class="label-dark"
        >
          Type: 
        </strong>
        <div
          class="detail-list"
        >
          <span>
            Grass /
          </span>
          <span>
            Poison
          </span>
        </div>
      </div>
      <div>
        <strong
          class="label-dark"
        >
          Resistant: 
        </strong>
        <div
          class="detail-list"
        >
          <span>
            Water /
          </span>
          <span>
            Electric /
          </span>
          <span>
            Grass /
          </span>
          <span>
            Fighting /
          </span>
          <span>
            Fairy
          </span>
        </div>
      </div>
      <div>
        <strong
          class="label-dark"
        >
          weaknesses: 
        </strong>
        <div
          class="detail-list"
        >
          <span>
            Fire /
          </span>
          <span>
            Ice /
          </span>
          <span>
            Flying /
          </span>
          <span>
            Psychic
          </span>
        </div>
      </div>
    </div>
    <div
      class="card detail"
      data-testid="pokemon-attacks"
    >
      <h3
        class="label-dark"
      >
        Attacks
      </h3>
      <div>
        <strong
          class="label-dark"
        >
          Fast Attacks: 
        </strong>
        <div
          class="detail-object-list"
        >
          <span>
            <strong
              class="label-light"
            >
              Name:
            </strong>
             Tackle
          </span>
          <span>
            <strong
              class="label-light"
            >
              Type:
            </strong>
             Normal
          </span>
          <span>
            <strong
              class="label-light"
            >
              Damage:
            </strong>
             12
          </span>
        </div>
        <div
          class="detail-object-list"
        >
          <span>
            <strong
              class="label-light"
            >
              Name:
            </strong>
             Vine Whip
          </span>
          <span>
            <strong
              class="label-light"
            >
              Type:
            </strong>
             Grass
          </span>
          <span>
            <strong
              class="label-light"
            >
              Damage:
            </strong>
             7
          </span>
        </div>
      </div>
      <div>
        <strong
          class="label-dark"
        >
          Special Attacks: 
        </strong>
        <div
          class="detail-object-list"
        >
          <span>
            <strong
              class="label-light"
            >
              Name:
            </strong>
             Power Whip
          </span>
          <span>
            <strong
              class="label-light"
            >
              Type:
            </strong>
             Grass
          </span>
          <span>
            <strong
              class="label-light"
            >
              Damage:
            </strong>
             70
          </span>
        </div>
        <div
          class="detail-object-list"
        >
          <span>
            <strong
              class="label-light"
            >
              Name:
            </strong>
             Seed Bomb
          </span>
          <span>
            <strong
              class="label-light"
            >
              Type:
            </strong>
             Grass
          </span>
          <span>
            <strong
              class="label-light"
            >
              Damage:
            </strong>
             40
          </span>
        </div>
        <div
          class="detail-object-list"
        >
          <span>
            <strong
              class="label-light"
            >
              Name:
            </strong>
             Sludge Bomb
          </span>
          <span>
            <strong
              class="label-light"
            >
              Type:
            </strong>
             Poison
          </span>
          <span>
            <strong
              class="label-light"
            >
              Damage:
            </strong>
             55
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
